zebra/zebra_vxlan.c

```diff
@@ -9041,19 +9041,19 @@ void zebra_vxlan_advertise_all_vni(ZAPI_HANDLER_ARGS)
 	struct stream *s = NULL;
 	int advertise = 0;
 	enum vxlan_flood_control flood_ctrl;
+	struct zebra_vrf *zvrf_default = NULL;
 
-	if (zvrf_id(zvrf) != VRF_DEFAULT) {
-		zlog_debug("EVPN VNI Adv for non-default VRF %u",
-			   zvrf_id(zvrf));
+	zvrf_default = zebra_vrf_lookup_by_id(VRF_DEFAULT);
```

Comment:
1) Previously, advertised_all_vni is permitted only on default VRF. EVPN BGP session also permitted on default VRF. So advertised_all_vni is equivalent to enabling EVPN. The check in old code works well. With the new code how to make sure advertised_all_vni is done only on the VRF in which EVPN BGP session is established ? 
2). Old code is making sure that EVPN VNI adv is not enabled on non default VRF.  But I do not understand this new change. This code is checking whether default VRF is present or not. Why to check this in this function and at this point of time ?

Reply:
> 1. Previously, advertised_all_vni is permitted only on default VRF. EVPN BGP session also permitted on default VRF. So advertised_all_vni is equivalent to enabling EVPN. The check in old code works well. With the new code how to make sure advertised_all_vni is done only on the VRF in which EVPN BGP session is established ?

As a matter of fact, no such check was present: you could perfectly do the following in a non default VRF previously:
```
router bgp 3 vrf blue
  router bgp 1337 vrf vrf42
  bgp router-id 172.16.1.1
  neighbor 172.16.1.6 remote-as internal
  address-family l2vpn evpn
    neighbor 172.16.1.6 activate
```
... but the session would never establish.

As for ensuring the advertisement is done only for the VRF with the EVPN session, it is quite tricky. This is marked by the activation of a neighbor in the evpn safi, but you'd need to still be able to pass both configuration commands (`advertise-all-vni` and `neighbor xxx activate`) in any order.

This all because I kept the previous definition of "evpn is enabled", which is a VRF has `advertise-all-vni` configured. A better definition would be "the VRF has an EVPN neighbor", but I preferred to keep the old one to stay as close the the previous behavior regarding the default VRF.

> 2. Old code is making sure that EVPN VNI adv is not enabled on non default VRF.  But I do not understand this new change. This code is checking whether default VRF is present or not. Why to check this in this function and at this point of time ?

As for previously, this is a habit of mine of checking returned pointer. But as the default VRF always exists, this check is not needed. As before, it could very well be removed.

Reply:
I think my earlier comment is not clear. What I mean is, by using the function is_evpn_enabled() in earlier code, which checks advertise_all_vni flag only on default VRF, it is ensuring that this is set or used only when this flag is set in default VRF. Even though there is no CLI restriction to set this flag in other non default VRFs, it doesn't have any use or effect. 

Now my comment related to the case where advertise all vni is done in a VRF A, and BGP session for EVPN address family is actually established in VRF B ? How does this work ?

Reply:
Ooooh I see what you mean. The new code won't actually exit if called to enable evpn on a vrf while already enabled in another vrf. This is prevented on the bgpd side by the cli handler, but not here.

As for a mismatch between advertise and evpn session, nothing would happen.
- Only the bgp vrf with advertise-all-vni will be registered to zebra, and no data would go from bgpd to zebra (no session, i.e. no data source), and all data coming from zebra would be ignored by bgpd since it has no peers to send it to
- The vrf with the session is not registered to zebra, so zebra would not send in any data, and all data that could be sent by bgpd will be ignored by zebra because the vrf and the evpn vrf do not match

What do you think about replacing this stupid `if (zvrf_default)` with something like

```c
/* Mismatch between EVPN VRF and current VRF
   (should be prevented by bgpd's cli */
if (is_evpn_enabled() && !zvrf->advertise_all_vni)
        return;
```

Reply:
Done in 77b998f